src/segment_reader.rs

```diff
@@ -342,6 +348,157 @@ impl AsyncSegmentReaderImpl {
     }
 }
 
+/// Prefetches data using AsyncSegmentReader.
+///
+/// It maintains a buffer and prefetches data to fill that buffer in the background so that
+/// every read call can read from the local buffer instead of waiting for server response.
+pub(crate) struct PrefetchingAsyncSegmentReader {
+    buffer: VecDeque<SegmentReadCommand>,
+    reader: Arc<Box<dyn AsyncSegmentReader>>,
+    read_size: usize,
+    pub(crate) offset: i64,
+    end_of_segment: bool,
+    receiver: Option<oneshot::Receiver<Result<SegmentReadCommand, ReaderError>>>,
+    handle: Handle,
+}
+
+// maximum number of buffered reply
+const BUFFERED_REPLY_NUMBER: usize = 2;
+
+impl PrefetchingAsyncSegmentReader {
+    pub(crate) fn new(
+        handle: Handle,
+        reader: Arc<Box<dyn AsyncSegmentReader>>,
+        offset: i64,
+        buffer_size: usize,
+    ) -> Self {
+        let mut wrapper = PrefetchingAsyncSegmentReader {
+            buffer: VecDeque::with_capacity(BUFFERED_REPLY_NUMBER),
+            reader,
+            read_size: buffer_size / BUFFERED_REPLY_NUMBER,
+            offset,
+            handle,
+            end_of_segment: false,
+            receiver: None,
+        };
+        wrapper.issue_request_if_needed();
+        wrapper
+    }
+
+    /// Reads from buffered data.
+    ///
+    /// Note: bytes returned could be less than requested.
+    pub(crate) async fn read(&mut self, buf: &mut [u8]) -> StdResult<usize, ReaderError> {
+        self.fill_buffer_if_available()?;
+        self.issue_request_if_needed();
+
+        while self.buffer.is_empty() {
+            if self.end_of_segment {
+                // no more bytes could be read from the current offset
+                return Ok(0);
+            }
+            self.fill_buffer().await?;
+            self.issue_request_if_needed();
+        }
+
+        let mut need_to_read = buf.len();
+        let mut copy_offset = 0;
+        while let Some(mut cmd) = self.buffer.pop_back() {
```

Comment:
According the the docs for VecDeque:
> The "default" usage of this type as a queue is to use push_back to add to the queue, and pop_front to remove from the queue.

I don't know if this is actually any faster. But as this direction is the "default" we should prefer it all else being equal.